src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -47,54 +62,107 @@ bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
       resolve_data_type(*expression_type, [&](auto type) { value = _calculate_expression(type, expression); });
     }
 
-    // If the value is std::nullopt, then the expression could not be resolved at this point,
-    // e.g. because it is not an arithmetic operator.
-    if (value != std::nullopt &&
-        _replace_expression_in_parents(node, node->output_column_references()[column_id], *value)) {
-      // If we successfully replaced the occurrences of the expression in the parent tree,
-      // we can remove the column here.
-      _remove_column_from_projection(projection_node, column_id);
+    if (value != std::nullopt) {
+      column_reference_to_value_map.emplace(projection_node->output_column_references()[column_id], *value);
     }
   }
-
-  return _apply_to_children(node);
 }
 
-bool ConstantCalculationRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
-                                                             const LQPColumnReference& expression_column,
-                                                             const AllTypeVariant& value) {
-  auto parent_tree_changed = false;
-  for (auto parent : node->parents()) {
-    if (parent->type() != LQPNodeType::Predicate) {
-      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
-      continue;
-    }
+bool ConstantCalculationRule::_replace_column_references_in_tree(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::map<LQPColumnReference, AllTypeVariant>& column_reference_to_value_map) {
+  auto tree_changed = false;
 
-    auto predicate_node = std::static_pointer_cast<PredicateNode>(parent);
+  if (node->left_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->left_child(), column_reference_to_value_map);
+  }
+  if (node->right_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->right_child(), column_reference_to_value_map);
+  }
+
+  if (node->type() == LQPNodeType::Predicate) {
+    auto predicate_node = std::static_pointer_cast<PredicateNode>(node);
 
     // We look for a PredicateNode which has an LQPColumnReference as value,
     // referring to the column which contains the Expression we resolved before.
-    if (is_lqp_column_reference(predicate_node->value()) &&
-        boost::get<LQPColumnReference>(predicate_node->value()) == expression_column) {
-      // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
-      auto new_predicate_node = std::make_shared<PredicateNode>(predicate_node->column_reference(),
-                                                                predicate_node->predicate_condition(), value);
-      predicate_node->replace_with(new_predicate_node);
-      parent_tree_changed = true;
+    if (is_lqp_column_reference(predicate_node->value())) {
```

Comment:
What about `predicate_node->column_reference()`? Uncommon, but might be in column_reference_to_value_map as well - and since we can't put a Value into the left argument of the PredicateNode, this should probably stop the removal of the Projection that produces the column from being deleted.

Consider a query like `SELECT * FROM (SELECT 1 + 4 as a, b FROM table_a) WHERE a > 3;` - weird, but could happen.

Reply:
This will probably be solved by also the outcome of the discussion below. If we do not remove the columns in the ProjectionNodes anyway, we are fine, and if we do, this will be included in the solution for that.